### PR TITLE
Add secrets and registry support

### LINKED ISSUE #10: Allow authentication in multiple registries
# Motivation

At the moment the workflow only allows to push to azure registry. It could be of interest to allow other registries like GitHub or DockerHub.

## Assumptions

* Registry login is performed at the python level. 
* We'll use the provider SDKs to perform the docker login. 

# Acceptance criteria

- [ ] Multiple registries are allowed
- [ ] Team acceptance

### LINKED ISSUE #13: [build_images]  Add support for resolve secrets from a Azure KeyVault / Parameter Store
### Motivation

We would like to be able to resolve secrets, in the `build_images` workflow config file, so any secret you need to be used in the building of the final image, can be available to the build process, in the recommended way:

https://docs.docker.com/build/ci/github-actions/secrets/

https://www.youtube.com/watch?v=IzdWk6nA_Ho ❗ ❗ 
[OIDC con AWS](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-azure#updating-your-github-actions-workflow)
[OIDC con Azure](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-azure#updating-your-github-actions-workflow)

### Acceptance criteria

### PR DESCRIPTION
**Pull request by itasahobby**
closes #10
closes #13

### CODE REVIEW

**Comment by itasahobby** [comment]
In order to login locally to private registries docker-py was not enough since it does not store the result in the local configuration file. Therefore we have to manually overwrite it.

File: firestarter/workflows/build_images/auth.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        raise ValueError(f'Unknown provider: {provider}')
+
+
+def login_registry(registry, auth):
+    docker_cfg_path = f"{getenv('HOME')}/.docker/config.json"
+    if not path.exists(docker_cfg_path):
+        config = {}
+    else:
+        with open(docker_cfg_path, 'r') as f:
+            config = json.load(f)
```
Comment (alambike) [inline firestarter/workflows/build_images/auth.py:23]: I would put this methods in a factory class, as class methods,  something like :

`DockerRegistryAuthFactory.provider_from_str(provider=xxx) -> Provider`

Also I would put the `login_registry` method in the base class (no abstract class, so  no ABC heritage).
Also I would rename the Provider(s) with a more concrete name:
- DockerRegistryAuth
- AwsOidcDockerRegistryAuth
- AzureOidcDockerRegistryAuth
- GenericDockerRegistryAuth # by token

**Comment by itasahobby** [comment]
Another interesting detail to keep in mind. To test it out you can use it by just having a file `~/.aws/credentials` in your local environment.

File: firestarter/workflows/build_images/build_images.py [hunk 2]
```diff
@@ -160,16 +160,12 @@ def execute(self):
...
-                cli = get_default_cli()
-                success = cli.invoke(['acr', 'login', '--name', registry])
-                if success != 0:
-                    raise Exception('Login to the Azure Container Registry failed.')
+        
+        # Log in to the Azure Container Registry for each on-premises active in the configuration file
+        for key in self.on_premises:
+            print(self.config.images[key])
+            provider = DockerRegistryAuthFactory.provider_from_str(self.config.images[key].auth_strategy)
+            provider.login_registry(self.config.images[key].registry, provider.get_registry_auth())
```
Comment (alambike) [inline firestarter/workflows/build_images/build_images.py:168]: I would set the registry name as a property of the "Provider".

I would also put the `get_registry_auth()` as a required method to be defined  by the concrete provider, and called by the login_registry() method, something like:
```
provider.login_registry()
```

```python
def login_registry(self):
   self.get_registry_auth()
   ...
```

File: firestarter/workflows/build_images/providers.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    @property
+    def secrets(self):
+        return self._secrets
+
+    def resolve(self):
+        for key, value in self.secrets.items():
+            secret_manager = SecretProviderFactory.provider_from_str(value)
+            self.secrets[key] = secret_manager.get_secret()
+        return self.secrets
```
Comment (alambike) [inline firestarter/workflows/build_images/providers.py:132]: At least this classes should be in another file!

**Review by alambike — APPROVED** [review]

**Pull request merged**